runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -90,4 +113,161 @@ public void healNode(@Nonnull String endpointToHeal, int retry, @Nonnull Duratio
+            Map<String, CompletableFuture<NodeView>> futureMap = new HashMap<>();
+
+            // Ping all nodes asynchronously
+            for (String endpoint : allNodes) {
+                CompletableFuture<NodeView> cf = new CompletableFuture<>();
+                try {
+                    cf = l.getManagementClient(endpoint).sendHeartbeatRequest();
+                } catch (Exception e) {
+                    // Requesting the heartbeat can cause NetworkException if connection cannot
+                    // be established.
+                    cf.completeExceptionally(e);
+                }
+                futureMap.put(endpoint, cf);
+            }
+
+            // Accumulate all responses.
+            for (String endpoint : futureMap.keySet()) {
+                try {
+                    nodeViewMap.put(endpoint, CFUtils.getUninterruptibly(futureMap.get(endpoint),
+                            WrongEpochException.class));
+                    counters.computeIfPresent(endpoint, (s, count) -> count + 1);
+                } catch (WrongEpochException wee) {
+                    counters.computeIfPresent(endpoint, (s, count) -> count + 1);
+                } catch (Exception ignored) {
+                    // Ignore all exceptions.
+                }
+            }
+        }
+
+        // Using the peer views from the nodes to determine health of the cluster.
+        Set<String> peerResponsiveNodes = new HashSet<>(allNodes);
+        nodeViewMap.values().stream()
+                .map(NodeView::getNetworkMetrics)
+                // Get all peer connectivity maps from all node views.
+                .map(NetworkMetrics::getPeerConnectivityView)
+                .flatMap(stringBooleanMap -> stringBooleanMap.entrySet().stream())
+                // Filter out all unresponsive nodes from all the fault detectors.
+                .filter(entry -> !entry.getValue())
+                .map(Entry::getKey)
+                // Remove the unresponsive nodes from the set of all nodes.
+                .forEach(peerResponsiveNodes::remove);
+        layout.getUnresponsiveServers().forEach(peerResponsiveNodes::remove);
+
+        // Analyzes the responsive nodes as seen by the fault detectors to determine the health
+        // of the cluster.
+        ClusterStatus clusterStatus = getClusterHealth(layout, peerResponsiveNodes);
+
+        // List of all nodes which are responsive to this client.
+        // Client connectivity to the cluster.
+        List<String> responsiveEndpoints = counters.entrySet().stream()
+                // Only count nodes which have counter > 0
+                .filter(entry -> entry.getValue() > 0)
+                .map(Entry::getKey)
+                .collect(Collectors.toList());
+
+        return new ClusterStatusReport(layout, allNodes.stream()
+                .collect(Collectors.toMap(o -> o,
+                        t -> responsiveEndpoints.contains(t) ? NodeStatus.UP : NodeStatus.DOWN)),
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "List<String>" cannot contain a "T" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2175)

Reply:
Hmm, this seems to be an error (responsiveEndpoints seems to be a List<String> and allNodes seems to be a Set<String>... but it's hard to tell

Reply:
I changed responsiveEndpoints to a Set too to be consistent.